app/scripts/views/settings/two_step_authentication.js

```diff
@@ -0,0 +1,173 @@
+
+const View = FormView.extend({
+  template: Template,
+  className: 'two-step-authentication',
+  viewName: 'settings.two-step-authentication',
+
+  events: {
+    'click .show-code-link': preventDefaultThen('_showCode'),
+    'click .totp-cancel': preventDefaultThen('cancel'),
+    'click .totp-confirm-code': preventDefaultThen('confirmCode'),
+    'click .totp-create': preventDefaultThen('createToken'),
+    'click .totp-delete': preventDefaultThen('deleteToken'),
+    'click .totp-refresh': preventDefaultThen('refresh'),
+  },
+
+  _checkTokenExists() {
+    const account = this.getSignedInAccount();
+    return account.checkTotpTokenExists()
+      .then((result) => {
+        this._hasToken = result.exists;
+      });
+  },
+
+  _sanitizeCode(code) {
+    // Remove spaces and `-`
+    return code.replace(/[- ]*/g, '');
+  },
+
+  _showQrCode() {
+    this.$('#totp').removeClass('hidden');
+  },
+
+  _showStatus() {
+    this.$('.security-list').removeClass('hidden');
+    this.$('.totp-refresh').removeClass('hidden');
+    this.$('.totp-confirm-code').addClass('hidden');
+  },
+
+  _hideStatus() {
+    this.$('.security-list').addClass('hidden');
+    this.$('.totp-refresh').addClass('hidden');
+    this.$('.totp-confirm-code').removeClass('hidden');
+  },
+
+  _showCode() {
+    $('.manual-code').removeClass('hidden');
+    $('.show-code-link').addClass('hidden');
+  },
+
+  _getFormattedCode(code) {
+    // Insert spaces every 4 characters
+    return code.replace(/(\w{4})/g, '$1 ');
+  },
+
+  _isPanelEnabled() {
+    if (this.getSearchParam('showTwoStepAuthentication')) {
+      return true;
+    }
+    return false;
+  },
+
+  beforeRender() {
+    // This panel is currently behind a feature flag. Only show it
+    // when the query param `showTwoStepAuthentication` is specified.
+    if (! this._isPanelEnabled()) {
+      return this.remove();
+    } else {
+      return this._checkTokenExists();
+    }
+  },
+
+  initialize() {
+    this._hasToken = false;
+    this._statusVisible = true;
+  },
+
+  setInitialContext(context) {
+    context.set({
+      hasToken: this._hasToken,
+      isPanelOpen: this.isPanelOpen(),
+      statusVisible: this._statusVisible
+    });
+  },
+
+  cancel() {
+    return this.render()
+      .then(() => this.navigate('/settings'));
+  },
+
+  createToken() {
+    const account = this.getSignedInAccount();
+    return account.createTotpToken()
+      .then(result => {
+        $('.qr-image').attr('src', result.qrCodeUrl);
+        $('.code').html(this._getFormattedCode(result.secret));
+        this._showQrCode();
+        this._hideStatus();
+      });
+  },
+
+  deleteToken() {
+    const account = this.getSignedInAccount();
+    return account.deleteTotpToken()
+      .then(() => {
+        this.displaySuccess(t('Two-step authentication removed'), {
+          closePanel: true
+        });
+        return this.render()
+          .then(() => {
+            this.navigate('/settings');
+          });
```

Comment:
can this logic above not pyramid stack but continue flat with promises? 

I see a `render` then a navigate that also renders? Could we use something like this to signal the message to show https://github.com/mozilla/fxa-content-server/blob/08e931e8482e71dc8359d0d97fe3b779c89844d1/app/scripts/views/mixins/connect-another-device-mixin.js#L71 ?